ext/openssl/openssl.c

```diff
@@ -4228,180 +4228,280 @@ static EVP_PKEY *php_openssl_pkey_init_dh(zval *data, bool *is_private)
 #ifdef HAVE_EVP_PKEY_EC
 #if PHP_OPENSSL_API_VERSION < 0x30000
 static bool php_openssl_pkey_init_legacy_ec(EC_KEY *eckey, zval *data, bool *is_private) {
+	BIGNUM *p = NULL, *a = NULL, *b = NULL, *g_x = NULL, *g_y = NULL, *order = NULL, *cofactor = NULL;
+	BIGNUM *x = NULL, *y = NULL, *d = NULL;
 	EC_GROUP *group = NULL;
-	EC_POINT *pnt = NULL;
-	BIGNUM *d = NULL;
-	zval *bn;
-	zval *x;
-	zval *y;
+	EC_POINT *point_g = NULL;
+	EC_POINT *point_q = NULL;
+	BN_CTX *bctx = BN_CTX_new();
+	zval *curve_name_zv = zend_hash_str_find(Z_ARRVAL_P(data), "curve_name", sizeof("curve_name") - 1);
+	zval *generator_zv = zend_hash_str_find(Z_ARRVAL_P(data), "generator", sizeof("generator") - 1);
+	zval *seed_zv = zend_hash_str_find(Z_ARRVAL_P(data), "seed", sizeof("seed") - 1);
 
 	*is_private = false;
 
-	if ((bn = zend_hash_str_find(Z_ARRVAL_P(data), "curve_name", sizeof("curve_name") - 1)) != NULL &&
-			Z_TYPE_P(bn) == IS_STRING) {
-		int nid = OBJ_sn2nid(Z_STRVAL_P(bn));
-		if (nid != NID_undef) {
-			group = EC_GROUP_new_by_curve_name(nid);
-			if (!group) {
-				php_openssl_store_errors();
+	if(curve_name_zv && Z_TYPE_P(curve_name_zv) == IS_STRING && Z_STRLEN_P(curve_name_zv) > 0) {
+		int nid = OBJ_sn2nid(Z_STRVAL_P(curve_name_zv));
+		group = EC_GROUP_new_by_curve_name(nid);
+		if (nid == NID_undef || !group) {
+				php_error_docref(NULL, E_WARNING, "Unknown curve name");
 				goto clean_exit;
-			}
-			EC_GROUP_set_asn1_flag(group, OPENSSL_EC_NAMED_CURVE);
-			EC_GROUP_set_point_conversion_form(group, POINT_CONVERSION_UNCOMPRESSED);
-			if (!EC_KEY_set_group(eckey, group)) {
-				php_openssl_store_errors();
+		}
+
+		EC_GROUP_set_asn1_flag(group, OPENSSL_EC_NAMED_CURVE);
+	} else {
+		OPENSSL_PKEY_SET_BN(data, p);
+		OPENSSL_PKEY_SET_BN(data, a);
+		OPENSSL_PKEY_SET_BN(data, b);
+		OPENSSL_PKEY_SET_BN(data, order);
+		OPENSSL_PKEY_SET_BN(data, g_x);
+		OPENSSL_PKEY_SET_BN(data, g_y);
+		OPENSSL_PKEY_SET_BN(data, cofactor);
+
+		if (!(p && a && b && order)) {
+			php_error_docref(NULL, E_WARNING, "Missing params: need \"p\", \"a\", \"b\", \"order\" ");
```

Comment:
I think it should check if one of those parameters is set and if it's not, it should keep the warning that curve name is set otherwise it will be produced even if someone mistype curve_name...

Reply:
The logic here should be either selecting the elliptic curve name or the custom curve parameters. The priority is to generate the curve through the elliptic curve name since it is the standard way. If the curve name is entered incorrectly, an error should be thrown to terminate the program. If the curve name is not set, then the logic for generating the curve using custom curve parameters should be executed. However, in the custom parameter branch, it might be necessary to determine whether all the required parameters are not set or only some of them are not set. If all parameters are not set, prompt that the curve name is missing. If part parameters are not set, prompt according to the content of my code.